Atomic.md

```diff
@@ -0,0 +1,253 @@
+
+WritePlaceholderFile:
+
+-   Fopen() fails to create a file
+
+    -   This is retriable today as a second attempt should see the
+         IsEmpty bit on the parent directory and call WritePlaceholderFile
+         again. 
+
+-   Ftruncate fails to expand the file to the correct size (or is never
+     reached because of a crash)
+
+    -   We have an incomplete placeholder file in the virtualization root
+         that cannot be recovered. Future attempts of
+         WritePlaceholderFile will fail because fopen(path, "wbx")
+         won't let us overwrite the existing file.
+
+-   Memcpy'ing the providerID xattr fails (or is never reached because
+     of a crash)
+
+    -   We have an incomplete placeholder file in the virtualization root
+        that cannot be recovered. There are no flags set on this file.
+
+    -   The MirrorProvider doesn't currently use the providerID for
+         anything (although it should fail if it's empty)
+
+    -   VFSForGit will be unable to parse the placeholder version from
+         the providerID, OnGetFileStream will fail so hydrating this
+         file will fail. Retrying will not help.
+
+-   Memcpy'ing the contentID xattr fails (or is never reached because of
+     a crash)
+
+    -   We have an incomplete placeholder file in the virtualization root
+        that cannot be recovered. There are no flags set on this file.
+
+
+    -   The MirrorProvider doesn't use contentID so no issues will occur
+
+    -   VFSForGit will be unable to get the SHA of the object that it
+         needs to download and hydrating this file will fail. Retrying
+         will not help.
+
+-   We crash before reaching InitializeEmptyPlaceholder
+
+    -   Placeholder does not have flags for virtualization root and
+        empty.
+
+    -   This file will never be expanded (but it will be the right size due to ftruncate) as the kext will exit when we
+         do not see the InVirtualizationRoot bit.
+
+-   We crash between setting the bits for 'in virtualization root' and 'is empty'
+
+    -   This file will never be expanded as the kext will exit
+            when we do not see the IsEmpty bit
+
+-   Chmod() fails or is never reached due to a crash
+
+    -   The default mode should be 755 (unless a user has played with
+         umask, but this hole is out of scope for this design) so
+         hydration of the file should succeed. However, if the file had
+         a mode other than 755, tools that depend on it to be another
+         mode will fail. Recovery would require manual intervention by
+         chmodding this file.
+
+I believe that all of these problems can be addressed by moving to a
+model where we create placeholders in a temporary location, mark them up
+and then rename them into place.
+
+Proposed Solution:
+------------------
+
+This solution creates placeholders in a .staging directory within the
```

Comment:
I like the new `.staging` name

One implementation detail you'll very likely have to worry about is how to avoid inheriting owner/group/acl's from the `.staging` folder as you create your placeholder files. If a user decides to update the permissions/inheritance on this folder, we will need to detect and update that before we create placeholders inside it. We will also need some comprehensive tests covering these Mac-specific hydration concerns.